2480 - Setup public database migrations

sorja opened the pull request:
## developers note

To test this branch, create a new database `fra-db-test` inside the docker:

```bash
# terminal
docker exec -it fra-db bash

# inside docker container
time createdb -U frap frap-dev-test

# to rerun migrations / drop db:
time dropdb -U frap frap-dev-test #drop db
time createdb -U frap frap-dev-test #create db
```

Update the `.env` 

```config
PGDATABASE=frap-dev-test
```

- [x] Update `public.migration` created on auto populate
- [x] Admin user creation should be under tools 

Reviewing the PR:
It is suggested to hide the deleted files in the PR view
<img width="303" alt="Screenshot 2024-09-16 at 12 34 54" src="https://github.com/user-attachments/assets/421db453-9261-4f23-8511-455f5225481e">

sorja committed:
2480 - Add new command: migration-public:create (#3903)

* 2480 - Add new command: migration-public:create

Includes step creation

* remove transaction

* remove unnecessary jest file

* 2480 - Remove client from template

* 2480 - Sort migrations and use Promises.each

sorja committed:
2480 - init public schema via ddl migration step (#3905)

* 2480 - add command migration-public:run

* 2480 - fix ddl: migration.id serial

* 2480 - public/ddl: introduce getCreatePublicSchemaDDL

* 2480 - ddl step: init public schema

* 2480 - PublicSchema: country.config: not null, default {}

sorja committed:
2480 - init admin user (#3906)

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

sorja committed:
2480 - Exclude initial public migrations from future runs (#3907)

* 2480 - Exclude initial public migrations from future runs

* remove on conflict clause

minotogna commented:
@sorja can we name this pr as its issue ?

sorja committed:
Merge branch 'development' into 2480-setup-public-database-migrations

sorja committed:
2480 - Delete old migration system and migrations (#3908)

* 2480 - Exclude initial public migrations from future runs

* delete deprecated db

* remove deprecated migration system

* 2480 - truncate public.migrations

* remove on conflict clause

* remove references to old migration system

* update readme

* 2480 - test.js.yml disable test running

* add ts

* 2480 - remove db-migrate

---------

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

sorja committed:
2480 - createUser/createAdmin: create admin tool (#3918)

sorja commented:
@minotogna 

Is this fine to avoid running public-migrations in current instances?

```
src/test/migrations/steps/20240909101347-step-exclude-public-migrations.ts
```

Or should we merge this file to `dev` before this branch?

(Note: I will update it when I have all migrations in place that are needed for this main branch)

minotogna commented:
> @minotogna 
> 
> 
> 
> Is this fine to avoid running public-migrations in current instances?
> 
> 
> 
> ```
> 
> src/test/migrations/steps/20240909101347-step-exclude-public-migrations.ts
> 
> ```
> 
> 
> 
> Or should we merge this file to `dev` before this branch?
> 
> 
> 
> (Note: I will update it when I have all migrations in place that are needed for this main branch)

Not sure I understand.
On current instances initial step should not run and we should have a migration step .

sorja commented:
> > @minotogna
> > Is this fine to avoid running public-migrations in current instances?
> > ```
> > 
> > src/test/migrations/steps/20240909101347-step-exclude-public-migrations.ts
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Or should we merge this file to `dev` before this branch?
> > (Note: I will update it when I have all migrations in place that are needed for this main branch)
> 
> Not sure I understand. On current instances initial step should not run and we should have a migration step .

We discussed yesterday:
- Running migration on startup would run the public migrations before the steps

minotogna commented:
> > > @minotogna
> 
> > > Is this fine to avoid running public-migrations in current instances?
> 
> > > ```
> 
> > > 
> 
> > > src/test/migrations/steps/20240909101347-step-exclude-public-migrations.ts
> 
> > > ```
> 
> > > 
> 
> > > 
> 
> > >     
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >     
> 
> > >   
> 
> > > Or should we merge this file to `dev` before this branch?
> 
> > > (Note: I will update it when I have all migrations in place that are needed for this main branch)
> 
> > 
> 
> > Not sure I understand. On current instances initial step should not run and we should have a migration step .
> 
> 
> 
> We discussed yesterday:
> 
> - Running migration on startup would run the public migrations before the steps

I don't remember we said that , it we didn't understand each other then .
The migration steps are executed manually . We decide the order .
I think in current instances , the steps gets executed before the public migrations , otherwise this logic doesn't work.

sorja committed:
2480 - Fix integration tests (#3913)

* 2480 - Exclude initial public migrations from future runs

* delete deprecated db

* remove deprecated migration system

* 2480 - truncate public.migrations

* remove on conflict clause

* remove references to old migration system

* update readme

* 2480 - test.js.yml disable test running

* add ts

* 2480 - remove db-migrate

* 2480 - migration init settings table

* 2480 - migration init countries table (atlantis)

* 2480 - controller/assessment/create: remove transaction from DDL creation

* 2480 - getCreateSchemaDDL: add assessment enums

* 2480 - integration test: use atlantis countries for testing and fix typos

* 2480 - test.js.yml: re-enable tests

* 2480 - test.js.yml: run public migrations before tests

* shorten code

* correctly number atlantis countries

* 2480  migrate all countries

---------

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

sorja committed:
2480 - Run migrations on app start (#3921)

* 2480 - run migration on start

* 2480 - run migration on dev

* use &&

---------

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

minotogna committed:
Merge branch 'development' into 2480-setup-public-database-migrations

sorja commented on src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts:
```diff
@@ -0,0 +1,162 @@
+    create index if not exists idx_activity_log_cycle_message on ${schemaName}.activity_log using btree (cycle_uuid, message);
+    create index if not exists idx_activity_log_filtering on ${schemaName}.activity_log using btree (assessment_uuid, cycle_uuid, country_iso, message, time);
+
+    create table if not exists ${schemaName}.migration_steps (
```
Not sure if this should be here, or should the script check and create if not exists?

sorja commented on src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts:
```diff
@@ -0,0 +1,162 @@
+    );
+
+    -- This table is created separately, but left here for consistency
+    create table if not exists ${schemaName}.migrations (
```
Same as above

Not sure if this should be here, or should the script check and create if not exists?


They seem like helper tables that are not directly related to our application
(I would even introduce some separate schema or something to separate them more)

minotogna replied:
interesting. should we have a separate schema with all migrations ?

sorja replied:
For me, it would make sense, let me know

minotogna replied:
Ok @sorja , it is cleaner with the migrations schema .

With two tables atm : public and steps 

In the migrations (public) let's add a step that initializes the schema and tables .

In the step that runs on current instances , move steps and initialize the public table .

What do you think ?

sorja replied:
My question is, what would be logical place to create the `migrations.public` and `migration.step` tables?

When running the first migration of the migration script, we don't have this table on an empty database and it relies on it.

So I would suggest always running `create table if not exists` before running the script, as it is ran quite rarely in the end, so it doesn't matter to check

Let me know

sorja replied:
I am referring to this way of doing it:
https://github.com/openforis/fra-platform/pull/3904/files#diff-c0946e94fffdfd9287c7d54e5dd3cb24cfff140c35caef52ae54652767ceeff3R17

minotogna replied:
it  must run before any migrations and on any new instances only.ok?

sorja replied:
Yes, it is currently (this pr) implemented as so, I will update the schema and schema creation in the creation of table
👍

sorja commented on src/tools/migrations/public/index.ts:
```diff
@@ -0,0 +1,65 @@
+const executedMigrations: Array<string> = []
+
+// TODO: Move to migration step / schema initialisation
+const tableDDL = `
```
This is also related to previous comments

sorja committed:
2480 - Fix public.migrations column timestamp and delete old migration entries from table (#3922)

* 2480 - fix public.migrations run_on timestamp, set default now

* fix migration step typo

* test.js.yml: fix command order

first run build
then run migrations

* 2480 - Remove unnecessary migration steps

Remove unnecessary migration steps
Merge the content to step-exclude-public-migrations

---------

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

mergify[bot] committed:
Merge branch 'development' into 2480-setup-public-database-migrations

mergify[bot] committed:
Merge branch 'development' into 2480-setup-public-database-migrations

mergify[bot] committed:
Merge branch 'development' into 2480-setup-public-database-migrations

sorja committed:
2480 - Introduce migrations schema (#3937)

* initial commit for migrations schema

* 3914 - Initialise migrations.public table through migration-step

mergify[bot] committed:
Merge branch 'development' into 2480-setup-public-database-migrations
